### PR TITLE
Swap non-deterministic ops with possibly deterministic ones

### PR DESCRIPTION
**Pull request by dakshvar22**
**Proposed changes**:
- ...

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by dakshvar22** [comment]
/modeltest

```yml
dataset_branch: "nib-test"
include:
 - dataset: ["Customer 1", "Customer 1", "Customer 1", "Customer 1"]
   config: ["Sparse + DIET(seq) + ResponseSelector(t2t)"]
```

File: rasa/utils/tensorflow/layers.py [hunk 1]
```diff
@@ -1137,11 +1146,36 @@ def _compute_softmax_loss(
...
+        #         tf.ones_like(softmax_logits[..., :1], tf.float32),
+        #         tf.zeros_like(softmax_logits[..., 1:], tf.float32),
+        #     ],
+        #     axis=-1,
+        # )
+
+        # for numerical stability of softmax,
+        # see https://stackoverflow.com/a/49212689/3001665
+        input_logits = tf.nn.softmax(
+            softmax_logits - tf.math.reduce_max(softmax_logits, axis=-1, keepdims=True)
```
Comment (ka-bu) [inline rasa/utils/tensorflow/layers.py:1160]: shifting the logits should only be necessary if you implemented the softmax by hand

File: rasa/utils/tensorflow/layers.py [hunk 2]
```diff
@@ -1137,11 +1146,36 @@ def _compute_softmax_loss(
...
         )
-        return softmax_loss
+
+        loss = -tf.math.log(tf.math.maximum(input_logits[..., 0], 1e-6))
```
Comment (ka-bu) [inline rasa/utils/tensorflow/layers.py:1163]: Not sure what you tried here. We can...
1. plug in the softmax definition into the cross entropy, pull out the numerator, and let the log kill the one exp (nothing more that we can do there)
2. make use of shifting (same as in softmax) to get more stable results

which should yield something like
```suggestion
        logits = softmax_logits # because that name is just confusing
        shifted_logits = logits - tf.reduce_max(logits)
        cross_entropy = - logits[...,0] + tf.reduce_logsumexp(logits)
        loss = tf.reduce_sum( cross_entropy )
```
There's also`tf.nn.log_softmax` but that would give us more summands than we'd need.

**Review by ka-bu — CHANGES_REQUESTED** [review]

File: rasa/utils/tensorflow/layers.py (hunk 2 above)
Comment (dakshvar22) [reply rasa/utils/tensorflow/layers.py:1163]: I assume in your suggestion `logits` is non-softmax applied version of logits. In that case, the above is exactly the same as my version.

File: rasa/utils/tensorflow/layers.py (hunk 1 above)
Comment (dakshvar22) [reply rasa/utils/tensorflow/layers.py:1160]: This was an additional harmless safeguard because `tf.nn.softmax` is implemented as a C++ function in tensorflow which is not visible in the TF source code right now unless we install it from source using `bazel` (the python version uses it through `gen_nn_ops` module which contains all the compiled C++ code in python). I didn't want to get caught up in doing that right now, can do once we are sure that making these changes is useful.

File: rasa/utils/tensorflow/layers.py (hunk 2 above)
Comment (ka-bu) [reply rasa/utils/tensorflow/layers.py:1163]: Nope, cause 1) there's no exp and then log taken from the numerator (which is what I'd think ce from logits would be doing), 2) there's no evaluation of softmax-normalised entires that aren't used, 3) don't think those changes would be optimized away

Comment (dakshvar22) [reply rasa/utils/tensorflow/layers.py:1163]: > Nope

I meant mathematically its the same and the above version doesn't break on at least our datasets. I wanted to try it to test if our datasets can break it.

> cause 1) there's no exp and then log taken from the numerator

yeah that's reduced but the subraction of `tf.reduce_max(logits)` and `tf.maximum` would act as safeguards.
Surely we should take your version if we merge this.

> there's no evaluation of softmax-normalised entires that aren't used

Not sure what you mean

**Comment by dakshvar22** [comment]
Status: Even with the above changes, the model training is not deterministic. Will need to try a different approach to first determine what's the first point of non-determinism in the model architecture. Clearly its not `sparse.sparse_dense_matmul`.

File: rasa/utils/tensorflow/layers.py (hunk 2 above)
Comment (ka-bu) [reply rasa/utils/tensorflow/layers.py:1163]: It's not the same considering that we don't have infinite precision. What I mean by the unused entries is that tf.nn.softmax will return a tensor of the same shape as the input, while we only really need those entries corresponding to the positive examples.

Comment (dakshvar22) [reply rasa/utils/tensorflow/layers.py:1163]: > It's not the same considering that we don't have infinite precision

Okay, you just mean that we should reduce the number of ops, so that precision is not lost in those ops at least, right?

> What I mean by the unused entries is that tf.nn.softmax will return a tensor of the same shape as the input, while we only really need those entries corresponding to the positive examples.

Sure thing, I didn't mean to have this as the most stable / polished version 👍

**Comment by twerkmeister** [comment]
Closing this as it does not look like this line of work yielded improvements.
